Not working on Alpine Linux (libc::pthread_t != std::os::linux::raw::pthread_t)

On Alpine Linux, `pthread_t` type is mismatched between `libc` and `std`.

* (libc) `ThreadId` = `libc::pthread_t`'s type is `*mut libc::c_void`!
* (std) `std::os::unix::thread::JoinHandleExt::as_pthread_t`'s return type is `RawPthread` = `u64`.

## Error reconstruction

```sh
docker run --rm -it alpine:3.19  # Digest: x86_64 => sha256:6457d53fb065d6f250e1504b9bc42d5b6c65941d57532c072d929dd0628977d0

# In the docker container, we'll do the rest
apk add cargo git hwloc-dev eudev-dev

git clone "https://github.com/HadrienG2/hwlocality.git"
cd hwlocality
cargo test --all  # Error!
```

### Logs

```sh
   Compiling hwlocality-sys v0.4.0 (/hwlocality/hwlocality-sys)
   Compiling hwlocality v1.0.0-alpha.1 (/hwlocality)
error[E0277]: `*mut c_void` doesn't implement `std::fmt::Display`
   --> src/cpu/binding.rs:805:62
    |
805 |             Self::Thread(id) => format!("the thread with TID {id}"),
    |                                                              ^^^^ `*mut c_void` cannot be formatted with the default formatter
    |
    = help: the trait `std::fmt::Display` is not implemented for `*mut c_void`
    = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
    = note: this error originates in the macro `$crate::__export::format_args` which comes from the expansion of the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> src/lib.rs:288:9
    |
286 |     fn as_thread_id(&self) -> ThreadId {
    |                               -------- expected `*mut c_void` because of return type
287 |         use std::os::unix::thread::JoinHandleExt;
288 |         self.as_pthread_t()
    |         ^^^^^^^^^^^^^^^^^^^ expected `*mut c_void`, found `u64`
    |
    = note: expected raw pointer `*mut c_void`
                      found type `u64`

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `hwlocality` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error[E0277]: `*mut libc::c_void` doesn't implement `std::fmt::Display`
   --> src/cpu/binding.rs:805:62
    |
805 |             Self::Thread(id) => format!("the thread with TID {id}"),
    |                                                              ^^^^ `*mut libc::c_void` cannot be formatted with the default formatter
    |
    = help: the trait `std::fmt::Display` is not implemented for `*mut libc::c_void`
    = note: in format strings you may be able to use `{:?}` (or {:#?} for pretty-print) instead
    = note: this error originates in the macro `$crate::__export::format_args` which comes from the expansion of the macro `format` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> src/lib.rs:288:9
    |
286 |     fn as_thread_id(&self) -> ThreadId {
    |                               -------- expected `*mut libc::c_void` because of return type
287 |         use std::os::unix::thread::JoinHandleExt;
288 |         self.as_pthread_t()
    |         ^^^^^^^^^^^^^^^^^^^ expected `*mut c_void`, found `u64`
    |
    = note: expected raw pointer `*mut libc::c_void`
                      found type `u64`

error: could not compile `hwlocality` (lib test) due to 2 previous errors
```

## Expected Solution

If we cast the type `ThreadId` to `*mut libc::c_void`, then all examples would be failed because `ThreadId` is not **thread-safe**.

So I think the plausible option is: `type ThreadId = u64`.

--- a/hwlocality-sys/Cargo.toml
+++ b/hwlocality-sys/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "hwlocality-sys"
-version = "0.4.0"
+version = "0.4.1"
 authors.workspace = true
 edition.workspace = true
 rust-version.workspace = true

--- a/hwlocality-sys/src/lib.rs
+++ b/hwlocality-sys/src/lib.rs
@@ -244,7 +244,9 @@ struct IncompleteType {
 
 /// Thread identifier (OS-specific)
 ///
-/// This is `HANDLE` on Windows and `libc::pthread_t` on all other platforms.
+/// This is `HANDLE` on Windows and `libc::pthread_t` on most other platforms,
+/// except on musl where it must be hardcoded to `c_ulong` to [preserve
+/// sanity](https://elixir.bootlin.com/musl/v1.2.4/source/include/alltypes.h.in#L53).
 #[cfg(target_os = "windows")]
 #[cfg_attr(docsrs, doc(cfg(all())))]
 pub type hwloc_thread_t = windows_sys::Win32::Foundation::HANDLE;
@@ -258,10 +260,21 @@ pub type hwloc_pid_t = u32;
 
 /// Thread identifier (OS-specific)
 ///
-/// This is `HANDLE` on Windows and `libc::pthread_t` on all other platforms.
-#[cfg(not(target_os = "windows"))]
+/// This is `HANDLE` on Windows and `libc::pthread_t` on most other platforms,
+/// except on musl where it must be hardcoded to `c_ulong` to [preserve
+/// sanity](https://elixir.bootlin.com/musl/v1.2.4/source/include/alltypes.h.in#L53)
+#[cfg(not(any(target_os = "windows", target_env = "musl")))]
 #[cfg_attr(docsrs, doc(cfg(all())))]
 pub type hwloc_thread_t = libc::pthread_t;
+
+/// Thread identifier (OS-specific)
+///
+/// This is `HANDLE` on Windows and `libc::pthread_t` on most other platforms,
+/// except on musl where it must be hardcoded to `c_ulong` to [preserve
+/// sanity](https://elixir.bootlin.com/musl/v1.2.4/source/include/alltypes.h.in#L53)
+#[cfg(target_env = "musl")]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+pub type hwloc_thread_t = c_ulong;
 
 /// Process identifier (OS-specific)
 ///

